libraries/AP_HAL/system.h

```diff
@@ -16,5 +16,6 @@ uint32_t micros();
 uint32_t millis();
 uint64_t micros64();
 uint64_t millis64();
+void delay_microseconds(uint16_t t);
```

Comment:
this function is normaly defined in Scheduler.h, why do we need to have it hal directy ?

Reply:
in scheduler as method so it is not accessable from low-level code which can't include AP_HAL.h

Reply:
There is some subsystems like OSD which runs on top of HAL too but can't include anything from Ardupilot for own reasons, but want to use some function. Yes this not required and I can do it in HAL, but I think that it will be reasonably to appent delay_microseconds to public interface

Reply:
Yes it makes sens

Reply:
also it would be cool to make additional PRs with inlining of functor.h, OwnPtr.h from AP_HAL folder and num_gcs() and chan() in GCS_Copter.h. This inlining refuces size and increases speed.

Reply:
The inline is always just a hint to the compiler and the function being declared and defined in the header with a simple statement. If you need it to be inlined for some other reason, then you'd have to add the always_inline attribute. Not the case here.

Applying just the inline part to OwnPtr.h and functor.h the binaries produced are exactly the same:

```bash
$ arm-none-eabi-size build/aerofc-v1/bin/arducopter build/aerofc-v1/bin/arducopter.old
   text	   data	    bss	    dec	    hex	filename
 885088	   2076	  57268	 944432	  e6930	build/aerofc-v1/bin/arducopter
 885088	   2076	  57268	 944432	  e6930	build/aerofc-v1/bin/arducopter.old
```

```bash
$ diff -u <(arm-none-eabi-objdump -d build/aerofc-v1/bin/arducopter.old) <(arm-none-eabi-objdump -d build/aerofc-v1/bin/arducopter)
--- /dev/fd/63	2018-01-30 11:37:19.971496370 -0800
+++ /dev/fd/62	2018-01-30 11:37:19.971496370 -0800
@@ -1,5 +1,5 @@
 
-build/aerofc-v1/bin/arducopter.old:     file format elf32-littlearm
+build/aerofc-v1/bin/arducopter:     file format elf32-littlearm
 
 
 Disassembly of section .text:
```

This was the diff between the .old and the second version:

```diff
diff --git a/libraries/AP_HAL/utility/OwnPtr.h b/libraries/AP_HAL/utility/OwnPtr.h
index 85a8957744..accfa46143 100644
--- a/libraries/AP_HAL/utility/OwnPtr.h
+++ b/libraries/AP_HAL/utility/OwnPtr.h
@@ -90,10 +90,10 @@ public:
         return _ptr;
     }
 
-    T& operator*() const { return *_ptr; }
-    T *operator->() const { return _ptr; }
-    bool operator !() const { return !_ptr; }
-    explicit operator bool() const { return _ptr != nullptr; }
+    inline T& operator*() const { return *_ptr; }
+    inline T *operator->() const { return _ptr; }
+    inline bool operator !() const { return !_ptr; }
+    inline explicit operator bool() const { return _ptr != nullptr; }
 
 private:
     T *_ptr;
diff --git a/libraries/AP_HAL/utility/functor.h b/libraries/AP_HAL/utility/functor.h
index 5954700853..74ca820e78 100644
--- a/libraries/AP_HAL/utility/functor.h
+++ b/libraries/AP_HAL/utility/functor.h
@@ -49,7 +49,7 @@ public:
         : Functor(nullptr, nullptr) { }
 
     // Call the method on the obj this Functor is bound to
-    RetType operator()(Args... args) const
+    inline RetType operator()(Args... args) const
     {
         return _method(_obj, args...);
     }
@@ -62,13 +62,13 @@ public:
     }
 
     // Allow to check if there's a method set in the Functor
-    explicit operator bool() const
+    inline explicit operator bool() const
     {
         return _method != nullptr;
     }
 
     template<class T, RetType (T::*method)(Args...)>
-    static constexpr Functor bind(T *obj)
+    inline static constexpr Functor bind(T *obj)
     {
         return { obj, method_wrapper<T, method> };
     }
```

So I think you could just remove this part from your changes.

Reply:
this may be GCC version dependent, with 4.8.3 there was a difference. Yes "inline" is just a hint, but it costs nothing and in no way harms, so it's better with it than without it. IMHO